### PR TITLE
[Bugfix] Harden Voxtral encoder melspec and chunk handling

### PR DESCRIPTION
**Pull request by artuskg**
## Purpose
Harden Voxtral encoder runtime behavior for robustness:
- register `mel_filters` and `hann_window` as non-persistent buffers,
- compute STFT/log-mel in FP32 and cast back to model dtype,
- make chunking robust when input length is not divisible by chunk size by padding last chunk and trimming encoder outputs back to true conv output lengths.

## Test Plan
- `python3 -m py_compile vllm/model_executor/models/voxtral.py`
- 
```bash
.venv/bin/python - <<'PY'
from types import SimpleNamespace
import torch
import torch.nn as nn
from vllm.model_executor.models.voxtral import VoxtralEncoderModel

enc = VoxtralEncoderModel.__new__(VoxtralEncoderModel)
nn.Module.__init__(enc)
enc.dtype = torch.float32
enc.config = SimpleNamespace(max_source_positions=4)
enc.whisper_encoder = SimpleNamespace(
    conv1=nn.Conv1d(80, 80, kernel_size=3, stride=1, padding=1),
    conv2=nn.Conv1d(80, 80, kernel_size=3, stride=2, padding=1),
)
enc.compute_whisper_melspec = lambda x: x

stacked, chunks_per_example, conv_out_lens = enc.prepare_inputs_for_conv([torch.randn(80, 10)])
assert stacked.shape == (2, 80, 8)
assert chunks_per_example == [2]
assert conv_out_lens == [4, 1]
PY
```

## Test Result
- `python3 -m py_compile vllm/model_executor/models/voxtral.py` passed.
- Focused runtime sanity script passed, confirming padded-last-chunk path and trimmed output lengths (`[4, 1]`).

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/model_executor/models/voxtral.py [hunk 1]
```diff
@@ -811,46 +826,83 @@ def downsample_factor(self) -> int:
...
+                    )
+
+                chunked_features.append(chunk)
+                conv_out_lens.append(conv_out_len)
+
             chunks_per_example.append(len(chunks))
 
         # [total_num_chunks, num_mel_bins, chunk_size]
-        return torch.stack(chunked_features), chunks_per_example
+        return torch.stack(chunked_features), chunks_per_example, conv_out_lens
```
[the remark this answers is not in the document]
Comment (artuskg) [reply vllm/model_executor/models/voxtral.py]: Addressed in 9aef348ac: added empty-input guards in both `prepare_inputs_for_conv` and `forward` to avoid empty-stack/runtime crashes.

File: vllm/model_executor/models/voxtral.py [hunk 2]
```diff
@@ -767,14 +767,28 @@ def __init__(
...
+        self.register_buffer(
+            "mel_filters",
+            torch.tensor(mel_filters, dtype=torch.float32),
+            persistent=False,
+        )
+        self.register_buffer(
+            "hann_window",
+            torch.hann_window(
+                self.config.window_size,
+                periodic=True,
```
Comment (patrickvonplaten) [inline vllm/model_executor/models/voxtral.py]:
```suggestion
```

Think the default is True and would like to stick to PT's default values

File: vllm/model_executor/models/voxtral.py [hunk 3]
```diff
@@ -811,46 +826,93 @@ def downsample_factor(self) -> int:
...
+
             chunks_per_example.append(len(chunks))
 
+        if not chunked_features:
+            empty = torch.empty(
+                (0, self.config.num_mel_bins, self.chunk_size),
+                dtype=self.dtype,
+                device=self.whisper_encoder.conv1.weight.device,
+            )
+            return empty, chunks_per_example, conv_out_lens
```
Comment (patrickvonplaten) [inline vllm/model_executor/models/voxtral.py]: Why is this needed?

Comment (patrickvonplaten) [reply vllm/model_executor/models/voxtral.py]: It shouldn't be possible that chunked_features is empty because `audio_waveforms` should always be non-empty? Think it might be nicer to just add an assert that makes sure the we never pass empty audio to this function

**Review by patrickvonplaten — COMMENTED** [review]
Cool PR - left some comments. 

I'm not sure we want to add all that padding logic to make sure the chunking works correctly because that's already done in `mistral_common` here: https://github.com/mistralai/mistral-common/blob/58bd44e4f86715b1edb1c9fcc812cab9e2d99df6/src/mistral_common/tokens/tokenizers/audio.py#L239 which is called in VoxtralProcessorAdapter.__call__ -> why do you think we need this logic?

File: vllm/model_executor/models/voxtral.py (hunk 2 above)
Comment (artuskg) [reply vllm/model_executor/models/voxtral.py]: Good point. Updated in acdf0f3a7 to drop the explicit `periodic=True` argument and use the PyTorch default for `torch.hann_window`.

File: vllm/model_executor/models/voxtral.py (hunk 3 above)
Comment (artuskg) [reply vllm/model_executor/models/voxtral.py]: Addressed in acdf0f3a7. I removed the local chunk padding/trimming path and now rely on mistral_common padding. I also added an explicit `audio_waveforms must be non-empty` assert and a divisibility invariant check before splitting to keep failures explicit if assumptions are violated.

File: vllm/model_executor/models/voxtral.py [hunk 4]
```diff
@@ -816,30 +830,36 @@ def prepare_inputs_for_conv(
...
             for audio in audio_waveforms
         ]
 
         chunked_features: list[torch.Tensor] = []
         chunks_per_example: list[int] = []
-        for feature in input_features:
+        for feature in log_mel_features:
+            if feature.shape[-1] % self.chunk_size != 0:
+                raise ValueError(
+                    "Expected log-mel length to be divisible by chunk_size; "
```
Comment (patrickvonplaten) [inline vllm/model_executor/models/voxtral.py:863]:
```suggestion
                    "Expected log-mel length to be divisible by chunk_size, but got {feature.shape[-1]=} and {self.chunk_size=}"
```

**Review by patrickvonplaten — APPROVED** [review]
Left one small nit - thanks!
